static/app/views/explore/components/schemaHintsList.tsx

```diff
@@ -39,20 +49,59 @@ function SchemaHintsList({
     return [...sectionSortedTags, ...otherTags];
   }, [filterTags]);
 
-  // only show 8 tags for now until we have a better way to decide to display them
-  // TODO: use resize observer to dynamically show more/less tags
-  const first8Tags = useMemo(() => {
-    return filterTagsSorted.slice(0, 8);
-  }, [filterTagsSorted]);
+  const [visibleHints, setVisibleHints] = useState(filterTagsSorted);
+
+  useEffect(() => {
+    // debounce calculation to prevent 'flickering' when resizing
+    const calculateVisibleHints = debounce(() => {
+      if (!schemaHintsContainerRef.current) {
+        return;
+      }
+
+      const containerWidth = schemaHintsContainerRef.current.clientWidth;
+      let currentWidth = 0;
+      const gap = 8;
+      const averageHintWidth = 250;
+      const seeFullListTagWidth = 150;
+
+      const visibleItems = filterTagsSorted.filter((_hint, index) => {
+        const element = schemaHintsContainerRef.current?.children[index] as HTMLElement;
+        if (!element) {
+          // add in a new hint if there is enough space (taking into account the 'See full list' tag)
+          if (containerWidth - seeFullListTagWidth - currentWidth >= averageHintWidth) {
+            currentWidth += averageHintWidth + (index > 0 ? gap : 0);
+            return true;
+          }
+          return false;
+        }
+
+        const itemWidth = element.offsetWidth;
+        currentWidth += itemWidth + (index > 0 ? gap : 0);
 
-  const tagHintsText = useMemo(() => {
-    return first8Tags.map(tag => `${tag?.key} is ...`);
-  }, [first8Tags]);
+        // the last item is the 'See full list' tag, so we need make sure it fits in the remaining space
+        return currentWidth <= containerWidth - seeFullListTagWidth;
+      });
+
+      setVisibleHints([...visibleItems, seeFullListTag]);
+    }, 100);
+
+    // initial calculation
+    calculateVisibleHints();
+
+    const resizeObserver = new ResizeObserver(calculateVisibleHints);
+    if (schemaHintsContainerRef.current) {
+      resizeObserver.observe(schemaHintsContainerRef.current);
+    }
+
+    return () => resizeObserver.disconnect();
+  }, [filterTagsSorted]);
 
   return (
-    <SchemaHintsContainer>
-      {tagHintsText.map(text => (
-        <SchemaHintOption key={text}>{text}</SchemaHintOption>
+    <SchemaHintsContainer ref={schemaHintsContainerRef}>
+      {visibleHints.map((hint, index) => (
+        <SchemaHintOption key={index} data-type={hint?.key}>
```

Comment:
Can we use the hint key as the key? Also, why would a hint be `undefined`/`null` if we're mapping over the visible hints? can we drop the optional accesses here?